### ISSUE TITLE
How to log extra field (baggage) using Logback?

### ISSUE THREAD

**Issue by asarkar** [open-telemetry/opentelemetry-java-instrumentation#1391]
Along with trace and span ids, I'd also like to log an internal request identifier for each log item. Please update the instructions [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/master/instrumentation/logback/logback-1.0.0/library) specifying whether this is possible, and if yes, how to do it.
Another option is to be able to use a custom trace id.

**Comment by trask** [comment]
hey @asarkar, can you give us a specific example of an internal request identifier that you would like to include in the log?

**Comment by asarkar** [comment]
@trask It's usually a Java [UUID](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/UUID.html). We have legacy systems that use a homegrown distributed tracing mechanism, and rely on this id being passed around. I'm using OTEL in a new project, but I need to be able to link to the calls made by the legacy system using the internal id.

For all intents and purposes, the internal id is equivalent to an OTEL trace id (and hence, if it's possible to use it as a custom trace id, that'll work just as well).

**Comment by trask** [comment]
are you looking to pass the UUID around via OpenTelemetry Context / Baggage, and then attach to MDC via something like a [scope listener](https://github.com/open-telemetry/opentelemetry-java/issues/922)?

**Comment by asarkar** [comment]
I'm not of scope listeners, but your description sounds about right. In plain English, I just want to see the internal id being logged for every line of log that logs OTEL trace and span.

> The only way to create an OpenTelemetry scope listener currently is by creating a class with a very special name io.grpc.override.ContextStorageOverride. This seems like a very odd public API surface that we are exposing

Your comment https://github.com/open-telemetry/opentelemetry-java/issues/922#issuecomment-661969038 makes me nervous; would you happen to have any examples if using the scope listener?

**Comment by trask** [comment]
Ya, we don't have that support yet, but working towards it.

@anuraaga this is a good context interop use case, being able to write a scope listener in application, and receiving callbacks when scope is changed by the javaagent

**Comment by anuraaga** [comment]
I think this sounds like it should be saved as a custom attribute on the Span and we should support populating Span attributes into logs in a configurable way.

Something Armeria has that I didn't get to yet ;)

**Comment by asarkar** [comment]
Do you’ve any examples/docs showing how to add additional fields to a trace/span? If I’m not wrong, those are called baggage.

**Comment by asarkar** [comment]
@anuraaga @iNikem Now that the 2 GA versions are out, can you comment on where this might fit in the roadmap?

**Comment by trask** [comment]
hey @asarkar, we may not have an ideal solution for this until the OpenTelemetry logs specification is completed.

In the meantime, can you write your own logback appender similar to [OpenTelemetryAppender.java](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/logback/logback-1.0/library/src/main/java/io/opentelemetry/instrumentation/logback/v1_0/OpenTelemetryAppender.java)?

**Comment by asarkar** [comment]
@trask The UUID is either sent by the client as a HTTP request header, or we generate it if not present. Looking at the code of `OpenTelemetryAppender`, it's not clear to me how such a class would get hold of the UUID. From the discussion above, I got the impression that adding custom attribute to a span is not supported.

**Comment by trask** [comment]
can you put the UUID into the OTel Context? then you can read it from the OTel context in your logback appender

**Comment by josephearl** [comment]
I have I think a similar need. Currently services at my organisation are using a custom `X-Some-Trace-ID` header. 

We'd like this header to be 1) added the logging context/MDC much like traceId is and 2) propagated to remote services. 

This would allow us to incrementally adopt OpenTelemetry.

Basically I'd like what Spring Cloud Sleuth provides with the `spring.sleuth.baggage.remote-fields` and `spring.sleuth.baggage.correlation-fields`.

**Comment by trask** [comment]
it seems reasonable to me to add baggage to the MDC (similar to trace_id)

this could be done for logback-mdc, log4j-mdc-1.2, and log4j-context-data instrumentation

unlike `trace_id`, `span_id` and `trace_flags` which have recommended names (https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/logs#trace-context-in-legacy-formats), I'm not sure what names we should expose these under

a couple of options:

* just the baggage key, e.g. `mykey`
* prefixing them with `baggage_`, e.g. `baggage_mykey`

**Comment by mateuszrzeszutek** [comment]
> if we're not sure about the naming, we can always hide this feature behind an experimental flag until there's more clarity

👍 

Does it make sense to add the baggage convention to the spec?

**Comment by trask** [comment]
Linking in some related reading material:
* https://github.com/open-telemetry/opentelemetry-specification/issues/867#issuecomment-696459504
* https://github.com/open-telemetry/opentelemetry-java/issues/4550#issuecomment-1168147670

**Comment by rjacobs91** [comment]
@trask if I were to use the appender approach how would I hook this in with an Otel Java agent?

**Comment by trask** [comment]
> @trask if I were to use the appender approach how would I hook this in with an Otel Java agent?

I'd probably recommend adding a `LogRecordProcessor` and adding the baggage to the `LogRecord` in `emit()`. You can add the `LogRecordProcessor` to the Java agent using the Java agent extension mechanism: https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension#extensions

**Comment by atulmaersk** [comment]
Hi @trask I also have a similar query, where we're using the otel java agent to generate the trace-id and span-id automatically in the logs,Now we want to have an additional field(baggage f.ex correlation-ID)  which we receive at the first service which need to be propagated to the downstream services and needs to be printed in the logs along with trace-id and span-id. How can we achieve this using the otel-api  ? Any code samples, recommendation will be helpful.

**Comment by trask** [comment]
Resolved by #7892
